### ISSUE TITLE
Add image explanation

### ISSUE THREAD

**Issue by thomasp85** [thomasp85/lime#35]
This should be the focus (beyond bug fixes) for the next version. It will bring the R version on par with the Python one.

One of the biggest challenge of this is the superpixel segmentation - a crude implementation in R can be seen [here](https://dahtah.wordpress.com/2017/03/24/superpixels-in-imager/)

Another possible challenge is general memory usage - image data is much larger so the permutations will take both time and space.

On top of my head I believe the input should be image files rather than in-memory images. We can then provide a preprocessor function as with text analysis to allow the user to get the image data into the format they need for the model. This will solve the memory issue of permutations, as well as the fact that there seem to be no common image class with widespread use in modeling in R.

I believe the magick package should provide all the infrastructure needed for the lime side of things...

@pommedeterresautee you are free to comment and suggest things for this - I'll take the implementation upon me.

**Comment by thomasp85** [comment]
Wow - the superpixel implementation is really slow, mainly due to k-means being slow with the required number of clusters. I'll try to play with some alternative clustering algorithms to see if it can be speed up...

**Comment by thomasp85** [comment]
Hmm... An alternative algorithm to SLIC is Grid Seams (http://ieeexplore.ieee.org/document/6816834/) - a C++ implementation is available for inspiration: https://github.com/richarddlu/grid_seams

**Comment by thomasp85** [comment]
Another possibility is to improve upon SLIC. The main bottleneck is the k-means algorithm, but it seems there's no need to consider all pixels simultaneously as we know that pixels belonging to the same cluster is located spatially close to each other...

**Comment by pommedeterresautee** [comment]
I have not tried anything applied to image and Lime, but currently I am playing with dbscan on another project. Would it mak things better? (perf are super good and there is only 1 reaured hyper parameter which can be the minimal size of a block, very handy)

**Comment by thomasp85** [comment]
Tried dbscan on a whim - crashed my r session for some reason... I’ll keep looking in to it

**Comment by thomasp85** [comment]
Got dbscan to work - not a good fit for this as the eps argument is not intuitive for image dimensions

**Comment by pommedeterresautee** [comment]
For Grid seams, have you tried with O3 flag?

**Comment by thomasp85** [comment]
Haven't tried Grid seams at all - currently implementing SLIC in Rcpp

**Comment by expectopatronum** [comment]
Hi @thomasp85, is there any way how I could contribute to this issue? (I'll also gladly create test cases or later on a demo notebook/tutorial since I have no experience with Rcpp...)

**Comment by thomasp85** [comment]
If you have an image classifier and some test images to share it would help greatly - I already have most of the framework set up but as I doesn’t really do any image classification myself it’s hard to make a proper test

**Comment by expectopatronum** [comment]
I'm thinking about training a classifier using `caret` (probably random forest) on the [fashion-mnist dataset](https://github.com/zalandoresearch/fashion-mnist). Is that ok?

**Comment by thomasp85** [comment]
I think we'll probably need some larger, more complex images? - have you seen [the article](https://arxiv.org/abs/1602.04938) and the use cases they have there?

But then again, I know next to nothing about image classification...

**Comment by expectopatronum** [comment]
Yes I saw it, but so far I couldn't figure out which dataset he was using. He also has an [example using MNIST data](https://github.com/marcotcr/lime/blob/master/doc/notebooks/Tutorial%20-%20MNIST%20and%20RF.ipynb). Do you know of any more interesting image datasets? Maybe the [CIFAR-10 dataset](http://www.cs.utoronto.ca/~kriz/cifar.html) is more interesting?

**Comment by expectopatronum** [comment]
I managed to train a `tensorflow` model on the CIFAR dataset, because with `caret` it never finished (I tried random forest and neural nets), maybe the data is too big. The model performance is not good yet but I was able to define `predict_model` and `model_type` in a way that it is recognized by `explain`. I'm pretty sure predict_model could/should be improved. But if you're interested I can show it to you and you can use the model for testing. Meanwhile I can try to improve the accuracy of the classifier.

**Comment by AdamSpannbauer** [comment]
Hi.  I installed the image branch to see how its current state would work with one of the pretrained models shipped with keras.

I ran into an error (which might be totally expected in the current state of the dev), and I thought I'd share my script here in case it aids the dev process at all.  The script I used is [here](https://github.com/AdamSpannbauer/r_keras_lime_image/blob/master/keras_vgg19_lime.R), where I attempted to use lime to explain predictions created by the pretrained VGG19 imagenet weights.

I created the 'image_explainer' object w/o error (I'm not sure how to verify if it was created correctly).  There was an error when calling `lime:::explain.imagefile`: `Error in ``*tmp*``[4, , ] : subscript out of bounds`.  The error occurs in [line 59 of image.R](https://github.com/thomasp85/lime/blob/image/R/image.R#L59); the first dimension value is `3` when running with the image_explainer created in my script.

Without exploration I changed the index in the line from `4` to `3` to see if it would work with dumb luck.  It then ran into error in [line 71](https://github.com/thomasp85/lime/blob/image/R/image.R#L71) where it seems as if the `perm` matrix dimensions are transposed from what they should be for the `colnames` assignment to work.

Not sure if any of this helps, but I'd be more than happy to help if there's any additional exploration needed for the dev of the image features in this version of lime.
